### ISSUE TITLE
Add GlobalFetch API - Headers, Request, Response

### ISSUE THREAD

**Issue by piuccio** [tmpvar/jsdom#1724]
https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API
https://developer.mozilla.org/en-US/docs/Web/API/Headers
https://developer.mozilla.org/en-US/docs/Web/API/Request
https://developer.mozilla.org/en-US/docs/Web/API/Response

**Comment by jods4** [comment]
+1 would be nice as General availability in browsers means more web apps use it and fail in JSDOM.
Given that JSDOM implements `XMLHttpRequest`, it could be as simple as running `whatwg-fetch` polyfill at startup, right?

**Comment by domenic** [comment]
No, unfortunately the whatwg-fetch polyfill is not very spec compliant. Someone needs to write an implementation of the fetch algorithm in the standard: https://fetch.spec.whatwg.org/#fetching

**Comment by jods4** [comment]
Yes, for now I'm using the whatwg polyfill in JSDOM.
I'm not sure how much "non-compliant" it is, but it seems to work well enough for my case.

It would be nice to see something built-in, though. Especially since the `Fetch` API has become commonplace.

**Comment by domenic** [comment]
Yeah, it definitely would be nice! I hope someone is able to pull-request an implementation of the spec.

**Comment by jhabdas** [comment]
Here's a tool which might help with the tests for the PR: https://www.npmjs.com/package/fetch-inject. Of course it's not unit tested because it requires this issue to be closed before it can be. 😉

**Comment by KyleGalvin** [comment]
My application tests run in node rather than the browser. The lack of server-side fetch API (and typescript bindings to match) has become a show-stopper.

This issue looks like the most promising solution.

Does anyone else know of a workaround for the mean-time? I can use fetch in the browser. I can use node-fetch for tests, but I can't switch on the two packages dynamically. One environment or the other will always fail.

**Comment by ghostd** [comment]
Is this issue still relevant since v15 supports `fetch`?

**Comment by domenic** [comment]
v15 does not support fetch.

**Comment by ghostd** [comment]
Sorry, i misread the Changelog about the 15.1, only Headers was provided

**Comment by thiscompsciguy** [comment]
+1 to this. fetch is natively [supported in nearly all modern browser](https://caniuse.com/#search=fetch)

**Comment by Dscheksn** [comment]
> Does anyone else know of a workaround for the mean-time? 

My workaround looks like this:

```
    const firstTodo: Todo = { title: 'Todo 1' };
    const secondTodo: Todo = { title: 'Todo 2' };
    const todos: Todo[] = [firstTodo, secondTodo];

    global['fetch'] = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(todos),
      })
    );

    await act(async () => {
      render(<App />, container);
    });
```

**Comment by alexfromapex** [comment]
![](http://www.quickmeme.com/img/0d/0d5acab0f8f65247f219f4fba3a2d8d7d240ddb11c3c4c5081ba01f6a055e600.jpg)

but in all seriousness, I have an easy workaround:

First, `npm install --save-dev whatwg-fetch`

.. then ..

jest.config.js:

    ...
      setupFiles: ['<rootDir>/setupJest.js'],
    ...

setupJest.js:

    ...
    import 'whatwg-fetch';
    ...


and then you'll automagically have all of the `fetch` related globals defined in your test code

**Comment by eweitz** [comment]
[Node 18 fully supports `fetch`](https://nodejs.org/ko/blog/announcements/v18-release-announce/) by default, and was released in April 2022.  Does that make this easier?

I ask because my use case would benefit from Cache API support in jsdom (#2422), which seems to depend on this.

**Comment by PetarKirov** [comment]
@domenic how feasible is for jsdom make `fetch` available when running on Node >= 18?

**Comment by toxeus** [comment]
@domenic I thought Node's fetch implementation is spec-compliant. Where does it deviate?

**Comment by SimenB** [comment]
Their readme has a bunch of examples of how they deviate: https://github.com/nodejs/undici/blob/main/README.md#undicifetchinput-init-promise

**Comment by Kikobeats** [comment]
is it a way to extend jsdom to add the missing APIs?

**Comment by SimenB** [comment]
There's an example in this very issue: https://github.com/jsdom/jsdom/issues/1724#issuecomment-720727999

**Comment by tkrotoff** [comment]
No more whatwg-fetch, node-fetch or undici thx to Node.js >= 18:

```TypeScript
// FixJSDOMEnvironment.ts

import JSDOMEnvironment from 'jest-environment-jsdom';

// https://github.com/facebook/jest/blob/v29.4.3/website/versioned_docs/version-29.4/Configuration.md#testenvironment-string
export default class FixJSDOMEnvironment extends JSDOMEnvironment {
  constructor(...args: ConstructorParameters<typeof JSDOMEnvironment>) {
    super(...args);

    // FIXME https://github.com/jsdom/jsdom/issues/3363
    this.global.structuredClone = structuredClone;

    // FIXME https://github.com/jsdom/jsdom/issues/1724
    this.global.fetch = fetch;
    this.global.Headers = Headers;
    this.global.Response = Response;
  }
}
```

```JavaScript
// jest.config.js

/** @type import('jest').Config */
const config = {
  testEnvironment: './FixJSDOMEnvironment.ts',
}

module.exports = config;
```

**Comment by philwolstenholme** [comment]
You might need to add `this.global.Request = Request;` in too if using @tkrotoff's suggestion.

**Comment by luchillo17** [comment]
I've fixed this by using the `import 'cross-fetch/polyfill';` on the test file (my library has a mixed environment, half of it is meant for node, the other half for jsdom, SSR and RSC in Next.js makes it hard to know what environment it should be using in tests).

**Comment by darrinmn9** [comment]
I just want to add on to the suggested workaround of patching `jest-environment-jsdom`, you might also need `AbortController` in order to abort fetch requests.

```
this.global.AbortController = AbortController
```

**Comment by osuritz-grammarly** [comment]
@tkrotoff I'm still getting this error (feels like a catch 22):
```
 ReferenceError: fetch is not defined

       7 |
       8 |     // FIXME https://github.com/jsdom/jsdom/issues/1724
    >  9 |     this.global.fetch = fetch;
```

**Comment by davidrapin** [comment]
If you are here and you need a quick workaround, this might work for you:
 ```ts
// ./jsdom-fetch.ts
import JSDOMEnvironment from 'jest-environment-jsdom';

export default class JSDOMFetch extends JSDOMEnvironment {
  constructor(...args) {
    super(...args);
    this.global.fetch = fetch;
    this.global.Request = Request;
    this.global.Response = Response;
  }
}
```

```ts
// ./jest.config.ts
 // [...]
 testEnvironment: './jsdom-fetch.ts',
 // [...]
```

solution based on https://stackoverflow.com/a/78051351/356323
